streams/src/main/java/org/apache/kafka/streams/state/internals/CachingKeyValueStore.java

```diff
@@ -83,10 +90,67 @@ public void init(final StateStoreContext context,
         streamThread = Thread.currentThread();
     }
 
-    Position getPosition() {
+    public Position getPosition() {
         return position;
     }
 
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public <R> QueryResult<R> query(final Query<R> query,
+                                    final PositionBound positionBound,
+                                    final QueryConfig config) {
+        if (config.isSkipCache()) {
+            return wrapped().query(query, positionBound, config);
+        }
+        final long start = config.isCollectExecutionInfo() ? System.nanoTime() : -1L;
+        QueryResult<R> result = null;
+        final int partition = context.taskId().partition();
+        final Position mergedPosition = Position.emptyPosition();
+        mergedPosition.merge(position);
+        mergedPosition.merge(wrapped().getPosition());
+        if (!StoreQueryUtils.isPermitted(mergedPosition, positionBound, partition)) {
+            result = QueryResult.notUpToBound(mergedPosition, positionBound, partition);
+        } else if (query instanceof KeyQuery) {
+            final KeyQuery<Bytes, byte[]> keyQuery = (KeyQuery<Bytes, byte[]>) query;
+            final Bytes key = keyQuery.getKey();
+
+            validateStoreOpen();
+            final Lock lock = this.lock.readLock();
+            lock.lock();
+            try {
+                validateStoreOpen();
+                if (context.cache() != null) {
+                    final LRUCacheEntry lruCacheEntry = context.cache().get(cacheName, key);
+                    if (lruCacheEntry != null) {
+                        final byte[] rawValue;
+                        if (timestampedSchema && !WrappedStateStore.isTimestamped(wrapped())) {
+                            rawValue = ValueAndTimestampDeserializer.rawValue(lruCacheEntry.value());
+                        } else {
+                            rawValue = lruCacheEntry.value();
+                        }
+                        result = (QueryResult<R>) QueryResult.forResult(rawValue);
+                    }
+                }
+            } finally {
+                lock.unlock();
+            }
+
+            if (result == null) {
+                result = wrapped().query(query, positionBound, config);
```

Comment:
I think we want to do this query unbounded, since the provided bound may include a position that hasn't been flushed yet. We've already validated the position above. Come to think of it, though, shouldn't we check the position after we grab the lock?